spinoso-symbol/src/inspect.rs

```diff
@@ -0,0 +1,734 @@
+use bstr::{ByteSlice, CharIndices};
+use core::iter::FusedIterator;
+use core::str::Chars;
+
+/// An iterator that yields a debug representation of a `Symbol` and its byte
+/// contents as a sequence of `char`s.
+///
+/// This struct is created by the [`inspect`] method on [`Symbol`]. See its
+/// documentation for more.
+///
+/// To format a `Symbol` directly into a writer, see [`format_inspect_into`].
+///
+/// [`inspect`]: crate::Symbol::inspect
+/// [`Symbol`]: crate::Symbol
+/// [`format_inspect_into`]: crate::Symbol::format_inspect_into
+#[must_use = "Iterator"]
+#[derive(Default, Debug, Clone)]
+#[cfg_attr(docsrs, doc(cfg(feature = "artichoke")))]
+pub struct Inspect<'a>(State<'a>);
+
+impl<'a> From<&'a str> for Inspect<'a> {
+    #[inline]
+    fn from(value: &'a str) -> Self {
+        Self::from(value.as_bytes())
+    }
+}
+
+impl<'a> From<&'a [u8]> for Inspect<'a> {
+    #[inline]
+    fn from(value: &'a [u8]) -> Self {
+        match value {
+            value if value.is_empty() => Self::default(),
+            // This match arm is known to be buggy. UTF-8 well-formedness is a
+            // necessary but not sufficient condition for having an unquoted
+            // debug representation. The byte contents must also be a valid Ruby
+            // identifier.
+            //
+            // See artichoke/artichoke#219.
+            value if value.is_utf8() => Self(State::unquoted(value)),
+            value => Self(State::quoted(value)),
+        }
+    }
+}
+
+impl<'a> Iterator for Inspect<'a> {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+
+impl<'a> DoubleEndedIterator for Inspect<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.0.next_back()
+    }
+}
+
+impl<'a> FusedIterator for Inspect<'a> {}
+
+#[must_use = "Iterator"]
+#[derive(Debug, Clone)]
+struct State<'a> {
+    string: &'a [u8],
+    sigil: bool,
+    open_quote: bool,
+    next: [Option<Literal>; 3],
+    iter: CharIndices<'a>,
+    close_quote: bool,
+}
```

Comment:
Implementing `Iterator` with all of this state would have been _much_ easier with generators.